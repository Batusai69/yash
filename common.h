/* Yash: yet another shell */
/* common.h: defines symbols common to all sources. */
/* (C) 2007-2010 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* This file should be included at the very first in every source file. */


#ifndef YASH_COMMON_H
#define YASH_COMMON_H

#if HAVE_CONFIG_H
# include "config.h"
#endif

#ifndef _POSIX_C_SOURCE
# define _POSIX_C_SOURCE 200112L
#endif
#ifndef _XOPEN_SOURCE
# define _XOPEN_SOURCE 600
#endif

#ifndef __GNUC__
# define __attribute__(ignore)
#endif

#ifdef __CYGWIN__
# undef __STRICT_ANSI__
#endif

#if HAVE_GETTEXT
# define gt(MSGID)  ((const char *) gettext(MSGID))
# define Ngt(MSGID) MSGID
# if HAVE_NGETTEXT
#  define ngt(MSGID,MSGS,MSGP,N) ((const char *) ngettext(MSGS, MSGP, N))
# else
#  define ngt(MSGID,MSGS,MSGP,N) gt(MSGID)
# endif
#else
# define gt(MSGID)              MSGID
# define Ngt(MSGID)             MSGID
# define ngt(MSGID,MSGS,MSGP,N) MSGID
#endif

#define ARGV(i) ((wchar_t *) argv[i])

#endif /* YASH_COMMON_H */


/* vim: set ts=8 sts=4 sw=4 noet tw=80: */
