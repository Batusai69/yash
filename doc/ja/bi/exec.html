<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
    "http://www.w3.org/TR/html4/strict.dtd">
<html lang="ja">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta http-equiv="Content-Script-Type" content="text/javascript">
<meta http-equiv="Content-Style-Type" content="text/css">
<meta name="author" content="magicant">
<meta name="description" content="Yash の exec 組込みコマンドについて">
<title>Yash マニュアル: exec 組込みコマンド</title>
<link rel="Contents" href="../index.html" title="目次">
<link rel="Prev" href="eval.html" title="Eval 組込みコマンド">
<link rel="Next" href="exit.html" title="Exit 組込みコマンド">
<link rel="Stylesheet" href="../style.css">

<div class="breadcrumb"><a href="../index.html">目次</a> >
<a rel="Up" href="index.html">組込みコマンド一覧</a></div>
<h1>Exec 組込みコマンド</h1>

<p><dfn>Exec 組込みコマンド</dfn>はシェルのプロセスを別の外部コマンドに置き換えます。またシェルのプロセスに対してリダイレクトを実行します。

<h2 id="syntax">構文</h2>
<ul>
<li><code>exec [-cf] [-a <var>コマンド名</var>] [<var>コマンド</var> [<var>引数</var>…]]</code>
</ul>

<h2 id="description">説明</h2>
<p>Exec コマンドを<var>コマンド</var>を指定して実行すると、シェルは<a href="../exec.html#simple">単純コマンドの実行</a>の最後のステップと同様にしてコマンドを実行します。ただし、コマンドは必ず外部コマンドとしてみなされ、関数や組込みコマンドは無視します。そしてその外部コマンドはサブシェルではなく現在の<a href="../exec.html#environment">コマンド実行環境</a>で exec システムコールを呼び出すことで実行します。これにより、シェルのプロセスは新しく起動するコマンドによって置き換わります。
<p>シェルが <a href="../posix.html">POSIX 準拠モード</a>のときまたは<a href="../interact.html">対話モード</a>でないとき、コマンドの起動に失敗するとシェルは直ちに終了します。
<p>シェルが POSIX 準拠モードではなくかつ対話モードのとき、コマンドの起動に失敗してもシェルは終了しません。
<p>シェルが POSIX 準拠モードではなくかつ対話モードのとき、停止中のジョブがあると、シェルは警告を表示し、コマンドを起動しません。警告を無視してコマンドを起動するには <code>-f</code> (<code>--force</code>) オプションを付けてください。
<p><var>コマンド</var>なしで実行した場合 exec コマンドは何も行いませんが、この exec コマンドを実行する際に行った<a href="../redir.html">リダイレクト</a>の効果は現在のコマンド実行環境に残ります。

<h2 id="options">オプション</h2>
<dl>
<dt><code>-a <var>コマンド名</var></code>, <code>--as=<var>コマンド名</var></code>
<dd><var>コマンド</var>の代わりに<var>コマンド名</var>をコマンド名としてコマンドに渡します。
<dt><code>-c</code>, <code>--clear</code>
<dd>既存の<a href="../params.html#variables">環境変数</a>をすべて削除した状態でコマンドを実行します。ただしこの exec コマンドを実行する際に行った変数代入の結果は環境変数としてコマンドに渡します。
<dt><code>-f</code>, <code>--force</code>
<dd>警告を無視してコマンドを実行します。
</dl>

<h2 id="operands">オペランド</h2>
<dl>
<dt><var>コマンド</var>
<dd>実行するコマンドです。
<dt><var>引数</var>…
<dd>実行するコマンドに渡すコマンドライン引数です。
</dl>

<h2 id="exitstatus">終了ステータス</h2>
<p>指定されたコマンドの起動に成功した場合、シェルのプロセスはそのコマンドのプロセスに置き換わってしまうので、終了ステータスはありません。
<p>実行しようとしたコマンドが見つからなかった場合、終了ステータスは 127 です。コマンドが見つかったが実行できなかった場合、終了ステータスは 126 です。コマンドを指定せずに exec コマンドを実行した場合、終了ステータスは 0 です。

<h2 id="notes">補足</h2>
<p>Exec コマンドは<a href="../builtin.html#types">特殊組込みコマンド</a>です。
<p>POSIX にはオプションに関する規定はありません。よってオプションは <a href="../posix.html">POSIX 準拠モード</a>では使えません。
