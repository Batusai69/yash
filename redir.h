/* Yash: yet another shell */
/* redir.h: manages file descriptors and provides functions for redirections */
/* (C) 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#ifndef REDIR_H
#define REDIR_H

#include <stdbool.h>
#include <stdio.h>

extern int xclose(int fd);
extern int xdup2(int oldfd, int newfd);

extern int ttyfd;

extern void init_shellfds(void);
extern void add_shellfd(int fd);
extern void remove_shellfd(int fd);
extern bool is_shellfd(int fd)
    __attribute__((pure));
extern void clear_shellfds(bool leavefds);
extern int copy_as_shellfd(int fd);
extern FILE *reopen_with_shellfd(FILE *f, const char *mode);
extern void open_ttyfd(void);
extern int get_ttyfd(void) __attribute__((pure));

typedef struct savefd_T savefd_T;
struct redir_T;

extern bool open_redirections(const struct redir_T *r, struct savefd_T **save);
extern void undo_redirections(savefd_T *save);
extern void clear_savefd(savefd_T *save);
extern void maybe_redirect_stdin_to_devnull(void);


#endif /* REDIR_H */


/* vim: set ts=8 sts=4 sw=4 noet: */
