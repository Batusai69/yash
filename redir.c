/* Yash: yet another shell */
/* redir.c: manages file descriptors and provides functions for redirections */
/* (C) 2007-2008 magicant */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#include "common.h"
#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <unistd.h>
#include "exec.h"
#include "expand.h"
#include "input.h"
#include "option.h"
#include "parser.h"
#include "path.h"
#include "redir.h"
#include "util.h"


/********** Utilities **********/

/* Closes a file descriptor surely.
 * If `close' returns EINTR, tries again.
 * If `close' returns EBADF, it is considered successful and silently ignored.
 * If `close' returns an error other than EINTR/EBADF, a message is printed. */
int xclose(int fd)
{
    while (close(fd) < 0) {
	switch (errno) {
	case EINTR:
	    continue;
	case EBADF:
	    return 0;
	default:
	    xerror(errno, Ngt("error in closing file descriptor %d"), fd);
	    return -1;
	}
    }
    return 0;
}

/* Performs `dup2' surely.
 * If `dup2' returns EINTR, tries again.
 * If `dup2' returns an error other than EINTR, an error message is printed.
 * `xclose' is called before `dup2'. */
int xdup2(int oldfd, int newfd)
{
    xclose(newfd);
    while (dup2(oldfd, newfd) < 0) {
	switch (errno) {
	case EINTR:
	    continue;
	default:
	    xerror(errno,
		    Ngt("cannot copy file descriptor %d to %d"),
		    oldfd, newfd);
	    return -1;
	}
    }
    return newfd;
}


/********** Shell FD **********/

static void reset_shellfdmin(void);


/* true iff stdin is redirected */
static bool is_stdin_redirected = false;

/* set of file descriptors used by the shell.
 * These file descriptors cannot be used by the user. */
static fd_set shellfds;
/* the minimum file descriptor that can be used for shell FD. */
static int shellfdmin;
/* the maximum file descriptor in `shellfds'.
 * `shellfdmax' is -1 if `shellfds' is empty. */
static int shellfdmax;

#ifndef SHELLFDMINMAX
#define SHELLFDMINMAX 100  /* maximum for `shellfdmin' */
#endif

/* file descriptor associated with the controlling terminal */
int ttyfd = -1;


/* Initializes shell FDs. */
void init_shellfds(void)
{
#ifndef NDEBUG
    static bool initialized = false;
    assert(!initialized);
    initialized = true;
#endif

    FD_ZERO(&shellfds);
    reset_shellfdmin();
    shellfdmax = -1;
}

/* Recomputes `shellfdmin'. */
void reset_shellfdmin(void)
{
    errno = 0;
    shellfdmin = sysconf(_SC_OPEN_MAX);
    if (shellfdmin == -1) {
	if (errno)
	    shellfdmin = 10;
	else
	    shellfdmin = SHELLFDMINMAX;
    } else {
	shellfdmin /= 2;
	if (shellfdmin > SHELLFDMINMAX)
	    shellfdmin = SHELLFDMINMAX;
	else if (shellfdmin < 10)
	    shellfdmin = 10;
    }
}

/* Adds the specified file descriptor (>= `shellfdmin') to `shellfds'. */
void add_shellfd(int fd)
{
    assert(fd >= shellfdmin);
    if (fd < FD_SETSIZE)
	FD_SET(fd, &shellfds);
    if (shellfdmax < fd)
	shellfdmax = fd;
}

/* Removes the specified file descriptor from `shellfds'.
 * Must be called BEFORE `xclose(fd)'. */
void remove_shellfd(int fd)
{
    if (0 <= fd && fd < FD_SETSIZE)
	FD_CLR(fd, &shellfds);
    if (fd == shellfdmax) {
	shellfdmax = fd - 1;
	while (shellfdmax >= 0 && !FD_ISSET(shellfdmax, &shellfds))
	    shellfdmax--;
    }
}
/* The argument to `FD_CLR' must be a valid (open) file descriptor. This is why
 * `remove_shellfd' must be called before closing the file descriptor. */

/* Checks if the specified file descriptor is in `shellfds'. */
bool is_shellfd(int fd)
{
    return fd >= FD_SETSIZE || (fd >= 0 && FD_ISSET(fd, &shellfds));
}

/* Closes all file descriptors in `shellfds' and empties it.
 * If `leavefds' is true, the file descriptors are actually not closed. */
void clear_shellfds(bool leavefds)
{
    if (!leavefds)
	for (int fd = 0; fd <= shellfdmax; fd++)
	    if (FD_ISSET(fd, &shellfds))
		xclose(fd);
    FD_ZERO(&shellfds);
    shellfdmax = -1;
    ttyfd = -1;
}

/* Duplicates the specified file descriptor as a new shell FD.
 * The new FD is added to `shellfds'.
 * On error, `errno' is set and -1 is returned. */
int copy_as_shellfd(int fd)
{
    int newfd = fcntl(fd, F_DUPFD, shellfdmin);
    if (newfd >= 0) {
	fcntl(newfd, F_SETFD, FD_CLOEXEC);
	add_shellfd(newfd);
    }
    return newfd;
}

/* Duplicates the underlining file descriptor of the specified stream.
 * The original stream is closed whether successful or not.
 * A new stream is open with the new FD using `fdopen' and returned.
 * The new stream's underlining file descriptor is registered as a shell FD and
 * set to non-blocking.
 * If NULL is given, this function just returns NULL without doing anything. */
FILE *reopen_with_shellfd(FILE *f, const char *mode)
{
    if (!f)
	return NULL;

    int newfd = copy_as_shellfd(fileno(f));
    fclose(f);
    if (newfd < 0 || !set_nonblocking(newfd))
	return NULL;
    else
	return fdopen(newfd, mode);
}

/* Opens `ttyfd' */
void open_ttyfd(void)
{
    if (ttyfd < 0) {
	int fd = open("/dev/tty", O_RDWR);
	if (fd >= 0) {
	    ttyfd = copy_as_shellfd(fd);
	    xclose(fd);
	}
	if (ttyfd < 0) {
	    xerror(errno, Ngt("cannot open `%s'"), "/dev/tty");
	    xerror(0, Ngt("job control disabled"));
	    do_job_control = false;
	}
    }
}


/********** Redirections **********/

/* info used to undo redirection */
struct savefd_T {
    struct savefd_T *next;
    int   sf_origfd;            /* original file descriptor */
    int   sf_copyfd;            /* copied file descriptor */
    bool  sf_stdin_redirected;  /* original `is_stdin_redirected' */
};

static void save_fd(int oldfd, savefd_T **save);
static int parse_and_check_dup(char *num, redirtype_T type)
    __attribute__((nonnull));


/* Opens a redirection.
 * If `save' is non-NULL, the original FD is saved and a pointer to the info is
 * assigned to `*save' (whether successful or not).
 * Returns true iff successful. */
bool open_redirections(const redir_T *r, savefd_T **save)
{
    if (save)
	*save = NULL;

    while (r) {
	if (r->rd_fd < 0 || is_shellfd(r->rd_fd)) {
	    xerror(0, Ngt("redirection: file descriptor %d unavailable"),
		    r->rd_fd);
	    return false;
	}

	/* expand rd_filename */
	char *filename;
	if (r->rd_type != RT_HERE && r->rd_type != RT_HERERT) {
	    filename = expand_single_with_glob(r->rd_filename, tt_single);
	    if (!filename)
		return false;
	} else {
	    filename = NULL;
	}

	/* save original FD */
	save_fd(r->rd_fd, save);

	/* now, open redirection */
	int fd;
	int flags;
	bool keepopen;
	switch (r->rd_type) {
	case RT_INPUT:
	    flags = O_RDONLY;
	    goto openwithflags;
	case RT_OUTPUT:
	    if (shopt_noclobber && !is_irregular_file(filename)) {
		flags = O_WRONLY | O_CREAT | O_EXCL;
	    } else {
	case RT_CLOBBER:
		flags = O_WRONLY | O_CREAT | O_TRUNC;
	    }
	    goto openwithflags;
	case RT_APPEND:
	    flags = O_WRONLY | O_CREAT | O_APPEND;
	    goto openwithflags;
	case RT_INOUT:
	    flags = O_RDWR | O_CREAT;
	    goto openwithflags;
openwithflags:
	    keepopen = false;
	    fd = open(filename, flags,
		    S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
	    if (fd < 0) {
		xerror(errno, Ngt("redirection: cannot open `%s'"), filename);
		free(filename);
		return false;
	    }
	    free(filename);
	    break;
	case RT_DUPIN:
	case RT_DUPOUT:
	    keepopen = true;
	    fd = parse_and_check_dup(filename, r->rd_type);
	    if (fd < -1)
		return false;
	    break;
	case RT_HERE:
	case RT_HERERT:
	    keepopen = false;
	    fd = open_heredocument(r->rd_herecontent);
	    if (fd < 0)
		return false;
	    break;
	default:
	    assert(false);
	}

	/* move the new FD to `r->rd_fd' */
	if (fd != r->rd_fd) {
	    if (fd >= 0) {
		if (xdup2(fd, r->rd_fd) < 0)
		    return false;
		if (!keepopen)
		    xclose(fd);
	    } else {
		xclose(r->rd_fd);
	    }
	}

	if (r->rd_fd == STDIN_FILENO)
	    is_stdin_redirected = true;

	r = r->next;
    }
    return true;
}

/* Saves the specified file descriptor if `save' is non-NULL. */
void save_fd(int fd, savefd_T **save)
{
    assert(0 <= fd);
    if (!save)
	return;

    int copyfd = copy_as_shellfd(fd);
    if (copyfd < 0 && errno != EBADF) {
	xerror(errno, Ngt("cannot save file descriptor %d"), fd);
	return;
    }

    savefd_T *s = xmalloc(sizeof *s);
    s->next = *save;
    s->sf_origfd = fd;
    s->sf_copyfd = copyfd;
    s->sf_stdin_redirected = is_stdin_redirected;
    /* note: if `fd' is formerly unused, `sf_copyfd' is -1. */
    *save = s;
}

/* Parses the argument to `RT_DUPIN'/`RT_DUPOUT'.
 * `num' is the argument to parse, which is expected to be "-" or numeric.
 * `num' is freed in this function.
 * If `num' is a positive integer, the value is returned.
 * If `num' is "-", -1 is returned.
 * Otherwise, a negative value other than -1 is returned. */
int parse_and_check_dup(char *const num, redirtype_T type)
{
    int fd = fd;
    if (strcmp(num, "-") == 0) {
	fd = -1;
    } else {
	char *end;
	errno = 0;
	long lfd = strtol(num, &end, 10);
	if (!isxdigit(num[0]) || *end != '\0')
	    errno = EINVAL;
	else if (lfd < 0)
	    errno = ERANGE;
	else
	    fd = (int) lfd;
	if (errno) {
	    xerror(errno, Ngt("redirection: %s"), num);
	    fd = -2;
	} else {
	    if (posixly_correct) {
		/* check the read/write permission */
		int flags = fcntl(fd, F_GETFL);
		if (flags < 0) {
		    xerror(errno, Ngt("redirection: %d"), fd);
		    fd = -2;
		} else {
		    if (type == RT_DUPIN && (flags & O_ACCMODE) == O_WRONLY) {
			xerror(0, Ngt("redirection: %d: not readable"), fd);
			fd = -2;
		    }
		    if (type == RT_DUPOUT && (flags & O_ACCMODE) == O_RDONLY) {
			xerror(0, Ngt("redirection: %d: not writable"), fd);
			fd = -2;
		    }
		}
	    }
	}
    }
    free(num);
    return fd;
}

/* Restores the saved file descriptor and frees `save'. */
void undo_redirections(savefd_T *save)
{
    while (save) {
	if (save->sf_copyfd >= 0) {
	    remove_shellfd(save->sf_copyfd);
	    xdup2(save->sf_copyfd, save->sf_origfd);
	    xclose(save->sf_copyfd);
	} else {
	    xclose(save->sf_origfd);
	}
	is_stdin_redirected = save->sf_stdin_redirected;

	savefd_T *next = save->next;
	free(save);
	save = next;
    }
}

/* Frees the FD-saving info without restoring FD.
 * The copied FDs are closed. */
void clear_savefd(savefd_T *save)
{
    while (save) {
	if (save->sf_copyfd >= 0) {
	    remove_shellfd(save->sf_copyfd);
	    xclose(save->sf_copyfd);
	}

	savefd_T *next = save->next;
	free(save);
	save = next;
    }
}

/* Redirects stdin to "/dev/null" if job control is off and stdin is not yet
 * redirected.
 * If `posixly_correct' is true, the condition is slightly different:
 * "if non-interactive" rather than "if job control is off". */
void maybe_redirect_stdin_to_devnull(void)
{
    int fd;

    if ((posixly_correct ? is_interactive : do_job_control)
	    || is_stdin_redirected)
	return;

    xclose(STDIN_FILENO);
    fd = open("/dev/null", O_RDONLY);
    if (fd > 0) {
	xdup2(fd, STDIN_FILENO);
	xclose(fd);
    }
    is_stdin_redirected = true;
}


/* vim: set ts=8 sts=4 sw=4 noet: */
