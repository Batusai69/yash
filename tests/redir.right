abc
abc
def
Point 1
../yash: $wp: parameter expanded to multiple words
Point 2
123
456
Point 3
Point 4
Point 5
789
Point 6
