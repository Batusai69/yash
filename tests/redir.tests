tmp=/tmp/yash-redir-test

echo abc > ${tmp}1
cat ${tmp}1

echo def >> ${tmp}1
cat < ${tmp}1

rm -f ${tmp}1
echo "Point 1"

# should be an error
wp="a b"
cat < $wp
echo "Point 2"

exec 3> ${tmp}2
exec 4> ${tmp}3
echo 123 1>&3
echo 456 1>&4
cat ${tmp}2 ${tmp}3
rm -f ${tmp}2 ${tmp}3
echo "Point 3"

exec 9>&-
echo "Point 4"

exec 5<>${tmp}4
echo 789 >&5
cat <&5
echo "Point 5"
cat ${tmp}4
exec 5<&-
rm -f ${tmp}4
echo "Point 6"

## various forms of null redirection
#test1 () {
#	if [ -f "$1" ]
#	then
#		rm -f "$1"
#	else
#		echo "$1" not found
#	fi
#}
#
#> /tmp/yash-null-redir-1
#test1 /tmp/yash-null-redir-1
#
#unset var
#$var > /tmp/yash-null-redir-2
#test1 /tmp/yash-null-redir-2
#
#( > /tmp/yash-null-redir-3 )
#test1 /tmp/yash-null-redir-3
#
#> /tmp/yash-null-redir-4 &
#wait
#test1 /tmp/yash-null-redir-4
#
## async command get /dev/null as stdin
#cat > /tmp/yash-null-redir-5 &
#wait
#test1 /tmp/yash-null-redir-5

