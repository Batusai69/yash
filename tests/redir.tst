echo ===== command redirection =====

cat <(echo 1)
cat <(echo 2)-
echo >(cat) 3 | cat

seq () {
	i=1
	while [ "$i" -le "$1" ]; do
		echo "$i"
		i=$((i+1))
	done
	unset i
}

seq 4 >(cat) | grep 4
grep 5 <(seq 5)
! $INVOKE $TESTEE -c '< (:)' 2>(cat>/dev/null)
echo $?
