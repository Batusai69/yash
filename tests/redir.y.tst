temp="${TESTTMP}/redir"

echo ===== pipe redirection =====

exec 3>&- 4>&- 5>&- 6>&-
exec 4>>|3; echo 4-3 >&4; exec 4>&-; cat <&3; exec 3<&-
exec 4>>|6; echo 4-6 >&4; exec 4>&-; cat <&6; exec 6<&-
exec 5>>|3; echo 5-3 >&5; exec 5>&-; cat <&3; exec 3<&-
exec 5>>|6; echo 5-6 >&5; exec 5>&-; cat <&6; exec 6<&-
exec 5>>|4; echo 5-4 >&5; exec 5>&-; cat <&4; exec 4<&-
exec 3>>|6; echo 3-6 >&3; exec 3>&-; cat <&6; exec 6<&-
exec 3>>|4; echo 3-4 >&3; exec 3>&-; cat <&4; exec 4<&-

(cat >"$temp" | echo loop) >>|0
(while read i; do if [ $i -lt 5 ]; then echo $((i+1)); else exit; fi done |
{ echo 0; tee -a "$temp"; }) >>|0
cat "$temp"

rm -f "$temp"


echo ===== command redirection =====

cat <(echo 1)
cat <(echo 2)-
echo >(cat) 3 | cat

seq () {
	i=1
	while [ "$i" -le "$1" ]; do
		echo "$i"
		i=$((i+1))
	done
	unset i
}

seq 4 >(cat) | grep 4
grep 5 <(seq 5)
! $INVOKE $TESTEE -c '< (:)' 2>(cat>/dev/null)
echo $?


echo ===== here-string =====

var=foo
cat <<<123
cat <<< "$var"
cat <<< "-
-"
