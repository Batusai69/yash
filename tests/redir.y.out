===== pipe redirection =====
4-3
4-6
5-3
5-6
5-4
3-6
3-4
loop
1
2
3
4
5
===== command redirection =====
1
2
3
4
5
0
